examples/zkp/cubic_circuit/main.go

```diff
@@ -0,0 +1,93 @@
+// Package cubic_circuit describes how to create and verify proofs on the Neo
+// blockchain. The example shows how to check that the prover knows the solution
+// of the cubic equation: y = x^3 + x + 5. The example is constructed for
+// BLS12-381 curve points using Groth-16 proving system. The example includes
+// everything that developer needs to start using ZKP on the Neo platform with
+// Go SDK:
+//  1. The described cubic circuit implementation.
+//  2. The off-chain proof generation with the help of gnark-crypto library.
+//  3. The Go verification contract generation and deployment with the help of
+//     NeoGo library.
+//  4. The on-chain proof verification for various sets of input data (implemented
+//     as end-to-end test).
+//  5. A set of unit-tests aimed to check the circuit validity.
+package cubic_circuit
+
+import (
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/cs/r1cs"
+)
+
+// CubicCircuit defines a simple circuit x**3 + x + 5 == y
+// that checks that the prover knows the solution of the provided expression.
+// The circuit must declare its public and secret inputs as frontend.Variable.
+// At compile time, frontend.Compile(...) recursively parses the struct fields
+// that contains frontend.Variable to build the frontend.constraintSystem.
+// By default, a frontend.Variable has the gnark:",secret" visibility.
+type CubicCircuit struct {
+	// Struct tags on a variable is optional.
+	// Default uses variable name and secret visibility.
+	X frontend.Variable `gnark:"x,secret"` // Secret input.
+	Y frontend.Variable `gnark:"y,public"` // Public input.
+}
+
+// A gnark circuit must implement the frontend.Circuit interface
+// (https://docs.gnark.consensys.net/HowTo/write/circuit_structure).
+var _ = frontend.Circuit(&CubicCircuit{})
+
+// Define declares the circuit constraints
+// x**3 + x + 5 == y.
+func (circuit *CubicCircuit) Define(api frontend.API) error {
+	x3 := api.Mul(circuit.X, circuit.X, circuit.X)
+
+	// Can be used for the circuit debugging.
+	api.Println("X^3", x3)
+
+	api.AssertIsEqual(circuit.Y, api.Add(x3, circuit.X, 5))
+	return nil
+}
+
+// main demonstrates how to build the proof and verify it with the help of gnark
+// library.
+func main() {
+	var (
+		circuit    CubicCircuit
+		assignment = CubicCircuit{X: 3, Y: 35}
+	)
+
+	// Compile our circuit into a R1CS (a constraint system).
+	ccs, _ := frontend.Compile(ecc.BLS12_381.ScalarField(), r1cs.NewBuilder, &circuit)
+
+	// Once the circuit is compiled, you can run the three algorithms of a zk-SNARK back end:
+
+	// 1. One time setup (groth16 zkSNARK).
+	pk, vk, _ := groth16.Setup(ccs)
```

Comment:
"Error handling omitted intentionally to simplify the example", worth mentioning somewhere.

Reply:
Fixed.